### PR TITLE
storage/factory: extend the Create method by newList and resourcePrefix params

### PR DESCRIPTION
**Pull request by p0lyn0mial**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
This PR extends the Create method of the storage factory to accommodate the new parameters. These new parameters will be utilised in the future (in https://github.com/kubernetes/kubernetes/pull/119557) by the etcd3 storage implementation to `GetCurrentResourceVersionFromStorage`
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
https://github.com/kubernetes/enhancements/issues/3157

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 1]
```diff
@@ -99,11 +99,11 @@ type objState struct {
 }
 
 // New returns an etcd3 implementation of storage.Interface.
-func New(c *clientv3.Client, codec runtime.Codec, newFunc func() runtime.Object, prefix string, groupResource schema.GroupResource, transformer value.Transformer, pagingEnabled bool, leaseManagerConfig LeaseManagerConfig) storage.Interface {
-	return newStore(c, codec, newFunc, prefix, groupResource, transformer, pagingEnabled, leaseManagerConfig)
+func New(c *clientv3.Client, codec runtime.Codec, newFunc, newListFunc func() runtime.Object, prefix, resourcePrefix string, groupResource schema.GroupResource, transformer value.Transformer, pagingEnabled bool, leaseManagerConfig LeaseManagerConfig) storage.Interface {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:102]: Can you clarify what those are needed for?

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:102]: Those params will be needed to call https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/util.go#L89

This method will give us an RV after which we are going to deliver our special bookmark event.
Does it make sense?

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:102]: Yes - thanks!

File: staging/src/k8s.io/apiserver/pkg/reconcilers/peer_endpoint_lease.go [hunk 2]
```diff
@@ -78,7 +78,9 @@ type peerEndpointLeaseReconciler struct {
 
 // NewPeerEndpointLeaseReconciler creates a new peer endpoint lease reconciler
 func NewPeerEndpointLeaseReconciler(config *storagebackend.ConfigForResource, baseKey string, leaseTime time.Duration) (PeerEndpointLeaseReconciler, error) {
-	leaseStorage, destroyFn, err := storagefactory.Create(*config, nil)
+	// note that newListFunc and resourcePrefix
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/reconcilers/peer_endpoint_lease.go:81]: What about newFunc ?

File: staging/src/k8s.io/apiserver/pkg/reconcilers/peer_endpoint_lease_test.go [hunk 3]
```diff
@@ -91,7 +91,7 @@ func TestPeerEndpointLeaseReconciler(t *testing.T) {
 	newFunc := func() runtime.Object { return &corev1.Endpoints{} }
 	sc.Codec = apitesting.TestStorageCodec(codecs, corev1.SchemeGroupVersion)
 
-	s, dFunc, err := factory.Create(*sc.ForResource(schema.GroupResource{Resource: "endpoints"}), newFunc)
+	s, dFunc, err := factory.Create(*sc.ForResource(schema.GroupResource{Resource: "endpoints"}), newFunc, nil, "")
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/reconcilers/peer_endpoint_lease_test.go:94]: Can we set newList at least in places where we set non-nil newFunc?

**Review by wojtek-t — COMMENTED** [review]
Just few minor comments now.

File: staging/src/k8s.io/apiserver/pkg/reconcilers/peer_endpoint_lease.go (hunk 2 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/reconcilers/peer_endpoint_lease.go:81]: Same, required for a watch req, you won't receive any bookmark events without a `newFunc`. I'll update the comment.

File: staging/src/k8s.io/apiserver/pkg/reconcilers/peer_endpoint_lease_test.go (hunk 3 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/reconcilers/peer_endpoint_lease_test.go:94]: sure, np.

**Pull request merged**